eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs

```diff
@@ -559,7 +561,43 @@ getCostModelParams :: CostModel -> [Integer]
 getCostModelParams (CostModel _ cm _) = cm
 
 decodeCostModelMap :: Decoder s (Map Language CostModel)
-decodeCostModelMap = decodeMapByKey fromCBOR decodeCostModel
+decodeCostModelMap =
+  ifDecoderVersionAtLeast
+    (natVersion @9)
+    (decodeMapByKey fromCBOR decodeCostModel)
+    (decodeMapByKey fromCBOR legacyDecodeCostModel)
+
+-- | Prior to version 9, each 'CostModel' was expected to be serialized as
+-- an array of integers of a specific length (depending on the version of Plutus).
+-- Starting in version 9, we allow the decoders to accept lists longer than what they
+-- require, so that new fields can be added in the future.
+-- For this reason, we must hard code the length expectation into the deserializers
+-- prior to version 9.
+--
+-- Note that the number of elements in the V1 and V2 cost models
+-- may change in the future, they are only fixed prior to version 9.
+--
+-- See https://github.com/input-output-hk/cardano-ledger/issues/2902
+legacyCostModelLength :: Language -> Int
+legacyCostModelLength PlutusV1 = 166
+legacyCostModelLength PlutusV2 = 175
+
+-- | See the note for 'legacyCostModelLength'.
+legacyDecodeCostModel :: Language -> Decoder s CostModel
+legacyDecodeCostModel lang = do
+  values <- fromCBOR
+  let numValues = length values
+      expectedNumValues = legacyCostModelLength lang
+  when (length values /= expectedNumValues) $
```

Comment:
A nitpick: could use `numValues` instead of `length values`

Reply:
:facepalm: that's what I meant to do! thank you for picking that nit before it became a lice infestation!!